manpages/brew.1
.\" generated by kramdown
.TH "BREW" "1" "April 2026" "Homebrew"
.SH NAME
brew \- The Missing Package Manager for macOS (or Linux)
.SH "SYNOPSIS"
\fBbrew\fP \fB\-\-version\fP
.br
\fBbrew\fP \fIcommand\fP [\fB\-\-verbose\fP|\fB\-v\fP] [\fIoptions\fP] [\fIformula\fP] \.\.\.
.SH "DESCRIPTION"
Homebrew is the easiest and most flexible way to install the UNIX tools Apple didn\[u2019]t include with macOS\. It can also install software not packaged for your Linux distribution without requiring \fBsudo\fP\&\.
.SH "TERMINOLOGY"
.TP
\fBformula\fP
Homebrew package definition that builds from upstream sources
.TP
\fBcask\fP
Homebrew package definition that installs pre\-compiled binaries built and signed by upstream
.TP
\fBprefix\fP
path in which Homebrew is installed, e\.g\. \fB/opt/homebrew\fP or \fB/home/linuxbrew/\.linuxbrew\fP
.TP
\fBkeg\fP
installation destination directory of a given \fBformula\fP version, e\.g\. \fB/opt/homebrew/Cellar/foo/0\.1\fP
.TP
\fBrack\fP
directory containing one or more versioned \fBkegs\fP, e\.g\. \fB/opt/homebrew/Cellar/foo\fP
.TP
\fBkeg\-only\fP
a \fBformula\fP is \fIkeg\-only\fP if it is not symlinked into Homebrew\[u2019]s prefix
.TP
\fBopt prefix\fP
a symlink to the active version of a \fBkeg\fP, e\.g\. \fB/opt/homebrew/opt/foo\fP
.TP
\fBCellar\fP
directory containing one or more named \fBracks\fP, e\.g\. \fB/opt/homebrew/Cellar\fP
.TP
\fBCaskroom\fP
directory containing one or more named \fBcasks\fP, e\.g\. \fB/opt/homebrew/Caskroom\fP
.TP
\fBexternal command\fP
\fBbrew\fP subcommand defined outside of the Homebrew/brew GitHub repository
.TP
\fBtap\fP
directory (and usually Git repository) of \fBformulae\fP, \fBcasks\fP and/or \fBexternal commands\fP
.TP
\fBbottle\fP
pre\-built \fBkeg\fP poured into a \fBrack\fP of the \fBCellar\fP instead of building from upstream sources
.SH "ESSENTIAL COMMANDS"
For the full command list, see the \fICOMMANDS\fP section\.
.P
With \fB\-\-verbose\fP or \fB\-\-debug\fP, many commands print extra debugging information\. Note that these options should only appear after a command\.
.P
Some command behaviour can be customised with environment variables; see the \fIENVIRONMENT\fP section\.
.SS "\fBinstall\fP \fIformula\fP"
Install \fIformula\fP\&\.
.P
\fIformula\fP is usually the name of the formula to install, but it has other syntaxes which are listed in the \fISPECIFYING FORMULAE\fP section\.
.SS "\fBuninstall\fP \fIformula\fP"
Uninstall \fIformula\fP\&\.
.SS "\fBlist\fP"
List all installed formulae\.
.SS "\fBsearch\fP \fR[\fItext\fP|\fB/\fP\fItext\fP\fB/\fP]"
Perform a substring search of cask tokens and formula names for \fItext\fP\&\. If \fItext\fP is flanked by slashes, it is interpreted as a regular expression\. The search for \fItext\fP is extended online to \fBhomebrew/core\fP and \fBhomebrew/cask\fP\&\. If no search term is provided, all locally available formulae are listed\.
.SH "COMMANDS"
.SS "\fBalias\fP \fR[\fB\-\-edit\fP] \fR[\fIalias\fP|\fIalias\fP=\fIcommand\fP]"
Show an alias\[u2019]s command\. If no alias is given, print the whole list\.
.TP
\fB\-\-edit\fP
Edit aliases in a text editor\. Either one or all aliases may be opened at once\. If the given alias doesn\[u2019]t exist it\[u2019]ll be pre\-populated with a template\.
.SS "\fBanalytics\fP \fR[\fIsubcommand\fP]"
Control Homebrew\[u2019]s anonymous aggregate user behaviour analytics\. Read more at 
.UR https://docs\.brew\.sh/Analytics
.UE \&\.
.TP
\fBbrew analytics\fP [\fBstate\fP]
Display the current state of Homebrew\[u2019]s analytics\.
.TP
\fBbrew analytics\fP (\fBon\fP|\fBoff\fP)
Turn Homebrew\[u2019]s analytics on or off respectively\.
.SS "\fBautoremove\fP \fR[\fB\-\-dry\-run\fP]"
Uninstall formulae that were only installed as a dependency of another formula and are now no longer needed\.
.TP
\fB\-n\fP, \fB\-\-dry\-run\fP
List what would be uninstalled, but do not actually uninstall anything\.
.SS "\fBbundle\fP \fR[\fIsubcommand\fP]"
Bundler for non\-Ruby dependencies from Homebrew, Homebrew Cask, Mac App Store dependencies, VSCode (and forks/variants) extensions, Go packages, Cargo packages, uv tools, Flatpak packages, Krew plugins and npm packages\.
.P
Note: Flatpak support is only available on Linux\.
.TP
\fBbrew bundle\fP [\fBinstall\fP]
Install and upgrade (by default) all dependencies from the \fBBrewfile\fP\&\.
.P
You can specify the \fBBrewfile\fP location using \fB\-\-file\fP or by setting the \fB$HOMEBREW_BUNDLE_FILE\fP environment variable\.
.P
You can skip the installation of dependencies by adding space\-separated values to one or more of the following environment variables: \fB$HOMEBREW_BUNDLE_BREW_SKIP\fP, \fB$HOMEBREW_BUNDLE_CASK_SKIP\fP, \fB$HOMEBREW_BUNDLE_MAS_SKIP\fP, \fB$HOMEBREW_BUNDLE_TAP_SKIP\fP\&\.
.TP
\fBbrew bundle upgrade\fP
Shorthand for \fBbrew bundle install \-\-upgrade\fP\&\.
.TP
\fBbrew bundle dump\fP
Write all installed casks/formulae/images/taps into a \fBBrewfile\fP in the current directory or to a custom file specified with the \fB\-\-file\fP option\.
.TP
\fBbrew bundle cleanup\fP
Uninstall all dependencies not present in the \fBBrewfile\fP\&\.
.P
This workflow is useful for maintainers or testers who regularly install lots of formulae\.
.P
Unless \fB\-\-force\fP is passed, this returns a 1 exit code if anything would be removed\.
.TP
\fBbrew bundle check\fP
Check if all dependencies present in the \fBBrewfile\fP are installed\.
.P
This provides a successful exit code if everything is up\-to\-date, making it useful for scripting\.
.TP
\fBbrew bundle list\fP
List all dependencies present in the \fBBrewfile\fP\&\.
.P
By default, only Homebrew formula dependencies are listed\.
.TP
\fBbrew bundle edit\fP
Edit the \fBBrewfile\fP in your editor\.
.TP
\fBbrew bundle add\fP \fIname\fP [\.\.\.]
Add entries to your \fBBrewfile\fP\&\. Adds formulae by default\. Use \fB\-\-cask\fP, \fB\-\-tap\fP, \fB\-\-vscode\fP, \fB\-\-go\fP, \fB\-\-cargo\fP, \fB\-\-uv\fP, \fB\-\-flatpak\fP, \fB\-\-krew\fP and \fB\-\-npm\fP to add the corresponding entry instead\.
.TP
\fBbrew bundle remove\fP \fIname\fP [\.\.\.]
Remove entries that match \fBname\fP from your \fBBrewfile\fP\&\. Use \fB\-\-formula\fP, \fB\-\-cask\fP, \fB\-\-tap\fP, \fB\-\-mas\fP, \fB\-\-vscode\fP, \fB\-\-go\fP, \fB\-\-cargo\fP, \fB\-\-uv\fP, \fB\-\-flatpak\fP, \fB\-\-krew\fP and \fB\-\-npm\fP to remove only entries of the corresponding type\. Passing \fB\-\-formula\fP also removes matches against formula aliases and old formula names\.
.TP
\fBbrew bundle exec\fP [\fB\-\-check\fP] [\fB\-\-no\-secrets\fP] \fIcommand\fP
Run an external command in an isolated build environment based on the \fBBrewfile\fP dependencies\.
.P
This sanitized build environment ignores unrequested dependencies, which makes sure that things you didn\[u2019]t specify in your \fBBrewfile\fP won\[u2019]t get picked up by commands like \fBbundle install\fP, \fBnpm install\fP, etc\. It will also add compiler flags which will help with finding keg\-only dependencies like \fBopenssl\fP, \fBicu4c\fP, etc\.
.TP
\fBbrew bundle sh\fP [\fB\-\-check\fP] [\fB\-\-no\-secrets\fP]
Run your shell in a \fBbrew bundle exec\fP environment\.
.TP
\fBbrew bundle env\fP [\fB\-\-check\fP] [\fB\-\-no\-secrets\fP]
Print the environment variables that would be set in a \fBbrew bundle exec\fP environment\.
.TP
\fB\-\-file\fP
Read from or write to the \fBBrewfile\fP from this location\. Use \fB\-\-file=\-\fP to pipe to stdin/stdout\.
.TP
\fB\-g\fP, \fB\-\-global\fP
Read from or write to the \fBBrewfile\fP from \fB$HOMEBREW_BUNDLE_FILE_GLOBAL\fP (if set), \fB${XDG_CONFIG_HOME}/homebrew/Brewfile\fP (if \fB$XDG_CONFIG_HOME\fP is set), \fB~/\.homebrew/Brewfile\fP or \fB~/\.Brewfile\fP otherwise\.
.TP
\fB\-v\fP, \fB\-\-verbose\fP
\fBinstall\fP prints output from commands as they are run\. \fBcheck\fP lists all missing dependencies\.
.TP
\fB\-\-no\-upgrade\fP
\fBinstall\fP does not run \fBbrew upgrade\fP on outdated dependencies\. \fBcheck\fP does not check for outdated dependencies\. Note they may still be upgraded by \fBbrew install\fP if needed\. Enabled by default if \fB$HOMEBREW_BUNDLE_NO_UPGRADE\fP is set\.
.TP
\fB\-\-upgrade\fP
\fBinstall\fP runs \fBbrew upgrade\fP on outdated dependencies, even if \fB$HOMEBREW_BUNDLE_NO_UPGRADE\fP is set\.
.TP
\fB\-\-upgrade\-formulae\fP
\fBinstall\fP runs \fBbrew upgrade\fP on any of these comma\-separated formulae, even if \fB$HOMEBREW_BUNDLE_NO_UPGRADE\fP is set\.
.TP
\fB\-\-install\fP
Run \fBinstall\fP before continuing to other operations, e\.g\. \fBexec\fP\&\.
.TP
\fB\-\-services\fP
Temporarily start services while running the \fBexec\fP or \fBsh\fP command\. Enabled by default if \fB$HOMEBREW_BUNDLE_SERVICES\fP is set\.
.TP
\fB\-f\fP, \fB\-\-force\fP
\fBinstall\fP runs with \fB\-\-force\fP/\fB\-\-overwrite\fP\&\. \fBdump\fP overwrites an existing \fBBrewfile\fP\&\. \fBcleanup\fP actually performs its cleanup operations\.
.TP
\fB\-\-cleanup\fP
\fBinstall\fP performs cleanup operation, same as running \fBcleanup \-\-force\fP\&\. Enabled by default if \fB$HOMEBREW_BUNDLE_INSTALL_CLEANUP\fP is set and \fB\-\-global\fP is passed\.
.TP
\fB\-\-all\fP
\fBlist\fP all dependencies\.
.TP
\fB\-\-formula\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP Homebrew formula dependencies\.
.TP
\fB\-\-cask\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP Homebrew cask dependencies\.
.TP
\fB\-\-tap\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP Homebrew tap dependencies\.
.TP
\fB\-\-mas\fP
\fBlist\fP or \fBdump\fP Mac App Store dependencies\.
.TP
\fB\-\-vscode\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP VSCode (and forks/variants) extensions\.
.TP
\fB\-\-go\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP Go packages\.
.TP
\fB\-\-cargo\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP Cargo packages\.
.TP
\fB\-\-uv\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP uv tools\.
.TP
\fB\-\-flatpak\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP Flatpak packages\. Note: Linux only\.
.TP
\fB\-\-krew\fP
\fBlist\fP or \fBdump\fP Krew plugins\.
.TP
\fB\-\-npm\fP
\fBlist\fP, \fBdump\fP or \fBcleanup\fP npm packages\.
.TP
\fB\-\-no\-vscode\fP
\fBdump\fP without VSCode (and forks/variants) extensions\. Enabled by default if \fB$HOMEBREW_BUNDLE_DUMP_NO_VSCODE\fP is set\.
.TP
\fB\-\-no\-go\fP
\fBdump\fP without Go packages\. Enabled by default if \fB$HOMEBREW_BUNDLE_DUMP_NO_GO\fP is set\.
.TP
\fB\-\-no\-cargo\fP
\fBdump\fP without Cargo packages\. Enabled by default if \fB$HOMEBREW_BUNDLE_DUMP_NO_CARGO\fP is set\.
.TP
\fB\-\-no\-uv\fP
\fBdump\fP without uv tools\. Enabled by default if \fB$HOMEBREW_BUNDLE_DUMP_NO_UV\fP is set\.
.TP
\fB\-\-no\-flatpak\fP
\fBdump\fP without Flatpak packages\. Enabled by default if \fB$HOMEBREW_BUNDLE_DUMP_NO_FLATPAK\fP is set\.
.TP
\fB\-\-no\-krew\fP
\fBdump\fP without Krew plugins\. Enabled by default if \fB$HOMEBREW_BUNDLE_DUMP_NO_KREW\fP is set\.
.TP
\fB\-\-no\-npm\fP
\fBdump\fP without npm packages\. Enabled by default if \fB$HOMEBREW_BUNDLE_DUMP_NO_NPM\fP is set\.
.TP
\fB\-\-describe\fP
\fBdump\fP and \fBadd\fP add a description comment above each line, unless the dependency does not have a description\. Enabled by default if \fB$HOMEBREW_BUNDLE_DESCRIBE\fP is set\.
.TP
\fB\-\-no\-restart\fP
\fBdump\fP does not add \fBrestart_service\fP to formula lines\.
.TP
\fB\-\-zap\fP
\fBcleanup\fP casks using the \fBzap\fP command instead of \fBuninstall\fP\&\.
.TP
\fB\-\-check\fP
Check that all dependencies in the Brewfile are installed before running \fBexec\fP, \fBsh\fP, or \fBenv\fP\&\. Enabled by default if \fB$HOMEBREW_BUNDLE_CHECK\fP is set\.
.TP
\fB\-\-no\-secrets\fP
Attempt to remove secrets from the environment before \fBexec\fP, \fBsh\fP, or \fBenv\fP\&\. Enabled by default if \fB$HOMEBREW_BUNDLE_NO_SECRETS\fP is set\.
.SS "\fBcasks\fP"
List all locally installable casks including short names\.
.SS "\fBcleanup\fP \fR[\fIoptions\fP] \fR[\fIformula\fP|\fIcask\fP \.\.\.]"
Remove stale lock files and outdated downloads for all formulae and casks, and remove old versions of installed formulae\. If arguments are specified, only do this for the given formulae and casks\. Removes all downloads more than 120 days old\. This can be adjusted with \fB$HOMEBREW_CLEANUP_MAX_AGE_DAYS\fP\&\.
.TP
\fB\-\-prune\fP
Remove all cache files older than specified \fIdays\fP\&\. If you want to remove everything, use \fB\-\-prune=all\fP\&\.
.TP
\fB\-n\fP, \fB\-\-dry\-run\fP
Show what would be removed, but do not actually remove anything\.
.TP
\fB\-s\fP, \fB\-\-scrub\fP
Scrub the cache, including downloads for even the latest versions\. Note that downloads for any installed formulae or casks will still not be deleted\. If you want to delete those too: \fBrm \-rf "$(brew \-\-cache)"\fP
.TP
\fB\-\-prune\-prefix\fP
Only prune the symlinks and directories from the prefix and remove no other files\.
.SS "\fBcommand\fP \fIcommand\fP \fR[\.\.\.]"
Display the path to the file being used when invoking \fBbrew\fP \fIcmd\fP\&\.
.SS "\fBcommand\-not\-found\-init\fP"
Print instructions for setting up the command\-not\-found hook for your shell\. If the output is not to a tty, print the appropriate handler script for your shell\.
.P
For more information, see: https://docs\.brew\.sh/Command\-Not\-Found
.SS "\fBcommands\fP \fR[\fB\-\-quiet\fP] \fR[\fB\-\-include\-aliases\fP]"
Show lists of built\-in and external commands\.
.TP
\fB\-q\fP, \fB\-\-quiet\fP
List only the names of commands without category headers\.
.TP
\fB\-\-include\-aliases\fP
Include aliases of internal commands\.
.SS "\fBcompletions\fP \fR[\fIsubcommand\fP]"
Control whether Homebrew automatically links external tap shell completion files\. Read more at 
.UR https://docs\.brew\.sh/Shell\-Completion
.UE \&\.
.TP
\fBbrew completions\fP [\fBstate\fP]
Display the current state of Homebrew\[u2019]s completions\.
.TP
\fBbrew completions\fP (\fBlink\fP|\fBunlink\fP)
Link or unlink Homebrew\[u2019]s completions\.
.SS "\fBconfig\fP, \fB\-\-config\fP"
Show Homebrew and system configuration info useful for debugging\. If you file a bug report, you will be required to provide this information\.
.SS "\fBdeps\fP \fR[\fIoptions\fP] \fR[\fIformula\fP|\fIcask\fP \.\.\.]"
Show dependencies for \fIformula\fP\&\. When given multiple formula arguments, show the intersection of dependencies for each formula\. By default, \fBdeps\fP shows all required and recommended dependencies\.
.P
If any version of each formula argument is installed and no other options are passed, this command displays their actual runtime dependencies (similar to \fBbrew linkage\fP), which may differ from a formula\[u2019]s declared dependencies\.
.P
\fINote:\fP \fB\-\-missing\fP and \fB\-\-skip\-recommended\fP have precedence over \fB\-\-include\-*\fP\&\.
.TP
\fB\-n\fP, \fB\-\-topological\fP
Sort dependencies in topological order\.
.TP
\fB\-1\fP, \fB\-\-direct\fP
Show only the direct dependencies declared in the formula\.

Library/Homebrew/bundle/brew.rb
# typed: true # rubocop:todo Sorbet/StrictSigil
# frozen_string_literal: true

require "json"
require "tsort"
require "utils/output"
require "bundle/package_type"

module Homebrew
  module Bundle
    class Brew < Homebrew::Bundle::PackageType
      extend Utils::Output::Mixin

      PACKAGE_TYPE = :brew
      PACKAGE_TYPE_NAME = "Formula"

      class << self
        def inherited(subclass)
          return if subclass.name == "Homebrew::Bundle::Brew::Services"

          super
        end

        def reset!
          require "bundle/brew_services"

          Homebrew::Bundle::Brew::Services.reset!
          @installed_formulae = nil
          @outdated_formulae = nil
          @pinned_formulae = nil
          @formulae = nil
          @formulae_by_full_name = nil
          @formulae_by_name = nil
          @formula_aliases = nil
          @formula_oldnames = nil
        end

        def preinstall!(name, no_upgrade: false, verbose: false, **options)
          new(name, options).preinstall!(no_upgrade:, verbose:)
        end

        def install!(name, preinstall: true, no_upgrade: false, verbose: false, force: false, **options)
          new(name, options).install!(preinstall:, no_upgrade:, verbose:, force:)
        end

        def install_verb(name, options = {})
          _ = options

          return "Installing" unless formula_upgradable?(name)

          "Upgrading"
        end

        def formula_installed_and_up_to_date?(formula, no_upgrade: false)
          return false unless formula_installed?(formula)
          return true if no_upgrade_with_args?(no_upgrade, formula)

          !formula_upgradable?(formula)
        end

        def no_upgrade_with_args?(no_upgrade, formula_name)
          no_upgrade && Bundle.upgrade_formulae.exclude?(formula_name)
        end

        def formula_in_array?(formula, array)
          return true if array.include?(formula)
          return true if array.include?(formula.split("/").last)

          old_name = formula_oldnames[formula]
          old_name ||= formula_oldnames[formula.split("/").last]
          return true if old_name && array.include?(old_name)

          resolved_full_name = formula_aliases[formula]
          return false unless resolved_full_name
          return true if array.include?(resolved_full_name)
          return true if array.include?(resolved_full_name.split("/").last)

          false
        end

        def formula_installed?(formula)
          formula_in_array?(formula, installed_formulae)
        end

        def formula_upgradable?(formula)
          # Check local cache first and then authoritative Homebrew source.
          formula_in_array?(formula, upgradable_formulae) && Formula[formula].outdated?
        end

        def installed_formulae
          @installed_formulae ||= formulae.map { |f| f[:name] }
        end

        def upgradable_formulae
          outdated_formulae - pinned_formulae
        end

        def outdated_formulae
          @outdated_formulae ||= formulae.filter_map { |f| f[:name] if f[:outdated?] }
        end

        def pinned_formulae
          @pinned_formulae ||= formulae.filter_map { |f| f[:name] if f[:pinned?] }
        end

        def formulae
          return @formulae if @formulae

          formulae_by_full_name
          @formulae
        end

        def formulae_by_full_name(name = nil)
          return @formulae_by_full_name[name] if name.present? && @formulae_by_full_name&.key?(name)

          require "formula"
          require "formulary"
          Formulary.enable_factory_cache!

          @formulae_by_name ||= {}
          @formulae_by_full_name ||= {}

          if name.nil?
            formulae = Formula.installed.map { add_formula(it) }
            sort!(formulae)
            return @formulae_by_full_name
          end

          formula = Formula[name]
          add_formula(formula)
        rescue FormulaUnavailableError => e
          opoo "'#{name}' formula is unreadable: #{e}"
          {}
        end

        def formulae_by_name(name)
          formulae_by_full_name(name) || @formulae_by_name[name]
        end

        def dump(describe: false, no_restart: false)
          require "bundle/brew_services"

          requested_formula = formulae.select do |f|
            f[:installed_on_request?] || !f[:installed_as_dependency?]
          end
          requested_formula.map do |f|
            brewline = if describe && f[:desc].present?
              f[:desc].split("\n").map { |s| "# #{s}\n" }.join
            else
              ""
            end
            brewline += "brew \"#{f[:full_name]}\""

            args = f[:args].map { |arg| "\"#{arg}\"" }.sort.join(", ")
            brewline += ", args: [#{args}]" unless f[:args].empty?
            brewline += ", restart_service: :changed" if !no_restart && Services.started?(f[:full_name])
            brewline += ", link: #{f[:link?]}" unless f[:link?].nil?
            brewline
          end.join("\n")
        end

        def dump_output(describe: false, no_restart: false)
          dump(describe:, no_restart:)
        end

        def fetchable_name(name, options = {}, no_upgrade: false)
          _ = options

          user, repository, = name.split("/", 3)
          return if user.present? && repository.present? &&
                    Homebrew::Bundle::Tap.installed_taps.exclude?("#{user}/#{repository}")
          return if formula_installed_and_up_to_date?(name, no_upgrade:)

          name
        end

        def formula_aliases
          return @formula_aliases if @formula_aliases

          @formula_aliases = {}
          formulae.each do |f|
            aliases = f[:aliases]
            next if aliases.blank?

            aliases.each do |a|
              @formula_aliases[a] = f[:full_name]
              if f[:full_name].include? "/" # tap formula
                tap_name = f[:full_name].rpartition("/").first
                @formula_aliases["#{tap_name}/#{a}"] = f[:full_name]
              end
            end
          end
          @formula_aliases
        end

        def formula_oldnames
          return @formula_oldnames if @formula_oldnames

          @formula_oldnames = {}
          formulae.each do |f|
            oldnames = f[:oldnames]
            next if oldnames.blank?

            oldnames.each do |oldname|
              @formula_oldnames[oldname] = f[:full_name]
              if f[:full_name].include? "/" # tap formula
                tap_name = f[:full_name].rpartition("/").first
                @formula_oldnames["#{tap_name}/#{oldname}"] = f[:full_name]
              end
            end
          end
          @formula_oldnames
        end

        private

        def add_formula(formula)
          hash = formula_to_hash formula

          @formulae_by_name[hash[:name]] = hash
          @formulae_by_full_name[hash[:full_name]] = hash

          hash
        end

        def formula_to_hash(formula)
          keg = if formula.linked?
            link = true if formula.keg_only?
            formula.linked_keg
          else
            link = false unless formula.keg_only?
            formula.any_installed_prefix
          end

          if keg
            require "tab"

            tab = Tab.for_keg(keg)
            args = tab.used_options.map(&:name)
            version = begin
              keg.realpath.basename
            rescue
              # silently handle broken symlinks
              nil
            end.to_s
            args << "HEAD" if version.start_with?("HEAD")
            installed_as_dependency = tab.installed_as_dependency
            installed_on_request = tab.installed_on_request
            runtime_dependencies = if (runtime_deps = tab.runtime_dependencies)
              runtime_deps.filter_map { |d| d["full_name"] }

            end
            poured_from_bottle = tab.poured_from_bottle
          end

          runtime_dependencies ||= formula.runtime_dependencies.map(&:name)

          bottled = if (stable = formula.stable) && stable.bottle_defined?
            bottle_hash = formula.bottle_hash.deep_symbolize_keys
            stable.bottled?
          end

          {
            name:                     formula.name,
            desc:                     formula.desc,
            oldnames:                 formula.oldnames,
            full_name:                formula.full_name,
            aliases:                  formula.aliases,
            any_version_installed?:   formula.any_version_installed?,
            args:                     Array(args).uniq,
            version:,
            installed_as_dependency?: installed_as_dependency || false,
            installed_on_request?:    installed_on_request || false,
            dependencies:             runtime_dependencies,
            build_dependencies:       formula.deps.select(&:build?).map(&:name).uniq,
            conflicts_with:           formula.conflicts.map(&:name),
            pinned?:                  formula.pinned? || false,
            outdated?:                formula.outdated? || false,
            link?:                    link,
            poured_from_bottle?:      poured_from_bottle || false,
            bottle:                   bottle_hash || false,
            bottled:                  bottled || false,
            official_tap:             formula.tap&.official? || false,
          }
        end

        def sort!(formulae)
          # Step 1: Sort by formula full name while putting tap formulae behind core formulae.
          #         So we can have a nicer output.
          formulae = formulae.sort do |a, b|
            if a[:full_name].exclude?("/") && b[:full_name].include?("/")
              -1
            elsif a[:full_name].include?("/") && b[:full_name].exclude?("/")
              1
            else
              a[:full_name] <=> b[:full_name]
            end
          end

          # Step 2: Sort by formula dependency topology.
          topo = Topo.new
          formulae.each do |f|
            topo[f[:name]] = topo[f[:full_name]] = f[:dependencies].filter_map do |dep|
              ff = formulae_by_name(dep)
              next if ff.blank?
              next unless ff[:any_version_installed?]

              ff[:full_name]
            end
          end
          @formulae = topo.tsort
                          .map { |name| @formulae_by_full_name[name] || @formulae_by_name[name] }
                          .uniq { |f| f[:full_name] }
        rescue TSort::Cyclic => e
          e.message =~ /\["([^"]*)".*"([^"]*)"\]/
          cycle_first = Regexp.last_match(1)
          cycle_last = Regexp.last_match(2)
          odie e.message if !cycle_first || !cycle_last

          odie <<~EOS
            Formulae dependency graph sorting failed (likely due to a circular dependency):
            #{cycle_first}: #{topo[cycle_first] if topo}
            #{cycle_last}: #{topo[cycle_last] if topo}
            Please run the following commands and try again:
              brew update
              brew uninstall --ignore-dependencies --force #{cycle_first} #{cycle_last}
              brew install #{cycle_first} #{cycle_last}
          EOS
        end
      end

      def initialize(name = "", options = {})
        super()
        @full_name = name
        @name = name.split("/").last
        @args = options.fetch(:args, []).map { |arg| "--#{arg}" }
        @conflicts_with_arg = options.fetch(:conflicts_with, [])
        @restart_service = options[:restart_service]
        @start_service = options.fetch(:start_service, @restart_service)
        @link = options.fetch(:link, nil)
        @postinstall = options.fetch(:postinstall, nil)
        @version_file = options.fetch(:version_file, nil)
        @changed = nil
      end

      def installed_and_up_to_date?(formula, no_upgrade: false)
        self.class.formula_installed_and_up_to_date?(formula, no_upgrade:)
      end

      def preinstall!(no_upgrade: false, verbose: false)
        if installed? && (self.class.no_upgrade_with_args?(no_upgrade, @name) || !upgradable?)
          puts "Skipping install of #{@name} formula. It is already installed." if verbose
          @changed = nil
          return false
        end

        true
      end

      def install!(preinstall: true, no_upgrade: false, verbose: false, force: false)
        install_result = if preinstall
          install_change_state!(no_upgrade:, verbose:, force:)
        else
          true
        end
        result = install_result

        if installed?
          service_result = service_change_state!(verbose:)
          result &&= service_result

          link_result = link_change_state!(verbose:)
          result &&= link_result

          postinstall_result = postinstall_change_state!(verbose:)
          result &&= postinstall_result

          if result && @version_file.present?
            # Use the version from the environment if it hasn't changed.
            # Strip the revision number because it's not part of the non-Homebrew version.
            version = if !changed? && (env_version = Bundle.formula_versions_from_env(@name))
              PkgVersion.parse(env_version).version
            else
              Formula[@full_name].version
            end.to_s
            File.write(@version_file, "#{version}\n")

            puts "Wrote #{@name} version #{version} to #{@version_file}" if verbose
          end
        end

        result
      end

      def install_change_state!(no_upgrade:, verbose:, force:)
        require "tap"
        if (tap_with_name = ::Tap.with_formula_name(@full_name))
          tap, = tap_with_name
          tap.ensure_installed!
        end

        return false unless resolve_conflicts!(verbose:)

        if installed?
          upgrade_formula!(verbose:, force:)
        else
          install_formula!(verbose:, force:)
        end
      end

      def start_service?
        @start_service.present?
      end

      def start_service_needed?
        require "bundle/brew_services"
        start_service? && !Services.started?(@full_name)
      end

      def restart_service?
        @restart_service.present?
      end

      def restart_service_needed?
        return false unless restart_service?

        # Restart if `restart_service: :always`, or if the formula was installed or upgraded
        @restart_service.to_s == "always" || changed?
      end

      def changed?
        @changed.present?
      end

      def service_change_state!(verbose:)
        require "bundle/brew_services"

        file = Services.versioned_service_file(@name)

        if restart_service_needed?
          puts "Restarting #{@name} service." if verbose
          Services.restart(@full_name, file:, verbose:)
        elsif start_service_needed?
          puts "Starting #{@name} service." if verbose
          Services.start(@full_name, file:, verbose:)
        else
          true
        end
      end

      def link_change_state!(verbose: false)
        link_args = []
        link_args << "--force" if unlinked_and_keg_only?

        cmd = case @link
        when :overwrite
          link_args << "--overwrite"
          "link" unless linked?
        when true
          "link" unless linked?
        when false
          "unlink" if linked?
        when nil
          if keg_only?
            "unlink" if linked?
          else
            "link" unless linked?
          end
        end

        if cmd.present?
          verb = "#{cmd}ing".capitalize
          with_args = " with #{link_args.join(" ")}" if link_args.present?
          puts "#{verb} #{@name} formula#{with_args}." if verbose
          return Bundle.brew(cmd, *link_args, @name, verbose:)
        end

        true
      end

      def postinstall_change_state!(verbose:)
        return true if @postinstall.blank?
        return true unless changed?

        puts "Running postinstall for #{@name}: #{@postinstall}" if verbose
        Kernel.system(@postinstall)
      end

      private

      def installed?
        self.class.formula_installed?(@name)
      end

      def linked?
        Formula[@full_name].linked?
      end

      def keg_only?
        Formula[@full_name].keg_only?
      end

      def unlinked_and_keg_only?
        !linked? && keg_only?
      end

      def upgradable?
        self.class.formula_upgradable?(@full_name)
      end

      def conflicts_with
        @conflicts_with ||= begin
          conflicts_with = Set.new
          conflicts_with += @conflicts_with_arg

          if (formula = self.class.formulae_by_full_name(@full_name)) &&
             (formula_conflicts_with = formula[:conflicts_with])
            conflicts_with += formula_conflicts_with
          end

          conflicts_with.to_a
        end
      end

      def resolve_conflicts!(verbose:)
        conflicts_with.each do |conflict|
          next unless self.class.formula_installed?(conflict)

          if verbose
            puts <<~EOS
              Unlinking #{conflict} formula.
              It is currently installed and conflicts with #{@name}.
            EOS
          end
          return false unless Bundle.brew("unlink", conflict, verbose:)

          next unless restart_service?

          require "bundle/brew_services"
          puts "Stopping #{conflict} service (if it is running)." if verbose
          Services.stop(conflict, verbose:)
        end

        true
      end

      def install_formula!(verbose:, force:)
        install_args = @args.dup
        install_args << "--force" << "--overwrite" if force
        install_args << "--skip-link" if @link == false
        with_args = " with #{install_args.join(" ")}" if install_args.present?
        puts "Installing #{@name} formula#{with_args}. It is not currently installed." if verbose
        unless Bundle.brew("install", "--formula", @full_name, *install_args, verbose:)
          @changed = nil
          return false
        end

        self.class.installed_formulae << @name
        @changed = true
        true
      end

      def upgrade_formula!(verbose:, force:)
        upgrade_args = []
        upgrade_args << "--force" if force
        with_args = " with #{upgrade_args.join(" ")}" if upgrade_args.present?
        puts "Upgrading #{@name} formula#{with_args}. It is installed but not up-to-date." if verbose
        unless Bundle.brew("upgrade", "--formula", @name, *upgrade_args, verbose:)
          @changed = nil
          return false
        end

        @changed = true
        true
      end

      class Topo < Hash
        include TSort

        def each_key(&block)
          keys.each(&block)
        end
        alias tsort_each_node each_key

        def tsort_each_child(node, &block)
          fetch(node.downcase).sort.each(&block)
        end
      end
    end
  end
end
